### PR TITLE
Remove the requirement for default configuration to be attached to the root tenant for each type of tenanted storage

### LINKED ISSUE #107: Remove the requirement for default configuration to be attached to the root tenant for each type of tenanted storage
For each of our tenant storage implementations (`Corvus.Azure.Cosmos.Tenancy`, `Corvus.Azure.Gremlin.Tenancy`, `Corvus.Azure.Storage.Tenancy` and `Corvus.Sql.Tenancy`) we provide extension methods on `IServiceCollection` to configure the appropriate factory classes.

All of these check that an appropriate `options` object has been supplied, and that this contains default configuration settings that will be applied to the root tenant.

We are now expecting that all configuration will be supplied from the tenant itself, and have determined that including configuration on the root tenant will, in most cases, be unnecessary.

### PR DESCRIPTION
**Pull request by jongeorge1**
Resolves #107 

At present, it's necessary to provide default configuration for the type of tenanted storage factory being used at the point it's added to the service container. For example, if adding tenanted Cosmos storage, you'd need to provide default Cosmos configuration to use when creating a tenanted container when no specific config has been provided.

This PR reverses that by removing the ability to provide default config. We now expect config to be added only if needed.

As such I've also modified the specs to add container-specific config for the test containers, as this is the approach we are expecting people to take when configuring storage.

One side effect of the change is that I have now required storage configuration to be supplied to the `TenantProviderBlobStoreServiceCollectionExtensions.AddTenantProviderBlobStore` method, requiring configuration for the root tenant's child tenant storage to be explicitly supplied, and rather than set this as the default storage config on the root tenant, it's being set as the tenancy container-specific config. This is also what is copied down to new child tenants, and this will avoid the problem of the tenancy service storage config being accidentally used by other services where storage config for their requirements has not been explicitly provided.

### CODE REVIEW

File: Solutions/Corvus.Azure.Storage.Tenancy/Microsoft/Extensions/DependencyInjection/TenancyBlobStorageServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -30,48 +29,33 @@ public static class TenancyBlobStorageServiceCollectionExtensions
...
         /// </summary>
-        /// <param name="services">The service collection.</param>
+        /// <param name="services">The target service collection.</param>
         /// <param name="getOptions">Function to get the configuration options.</param>
-        /// <returns>The modified service collection.</returns>
+        /// <returns>The service collection.</returns>
         public static IServiceCollection AddTenantCloudBlobContainerFactory(
-            this IServiceCollection services,
-            Func<IServiceProvider, TenantCloudBlobContainerFactoryOptions> getOptions)
+            this IServiceCollection services, Func<IServiceProvider,
```
Comment (idg10) [inline Solutions/Corvus.Azure.Storage.Tenancy/Microsoft/Extensions/DependencyInjection/TenancyBlobStorageServiceCollectionExtensions.cs]: You appaer to have split the type name `Func<IServierProvider, TenantCloudBlobContainerFactoryOptions>` across two lines! I think the newline should appear before the first `,` (i.e. before the `Func`).

File: Solutions/Corvus.Azure.Storage.Tenancy/Microsoft/Extensions/DependencyInjection/TenancyBlobStorageServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -30,48 +29,33 @@ public static class TenancyBlobStorageServiceCollectionExtensions
...
             if (services.Any(s => typeof(ITenantCloudBlobContainerFactory).IsAssignableFrom(s.ServiceType)))
             {
                 return services;
             }
 
             services.AddRootTenant();
-
-            services.AddTenantCloudBlobContainerFactory(
-                (sp, rootTenant) =>
+            services.AddTransient<BlobStorageConfiguration>();
```
Comment (idg10) [inline Solutions/Corvus.Azure.Storage.Tenancy/Microsoft/Extensions/DependencyInjection/TenancyBlobStorageServiceCollectionExtensions.cs]: This (and its equivalents on the other tenanted storage providers) confuses me. What's depending on this thing being available via DI? Aren't all `BlobStorageConfiguration` instances now for specific `BlobStorageContainerDefinition`s on specific tenants?

File: Solutions/Corvus.Tenancy.Storage.Azure.Blob/Corvus/Tenancy/Storage/Azure/TenantProviderBlobStore.cs [hunk 3]
```diff
@@ -166,14 +166,14 @@ public async Task<ITenant> CreateChildTenantAsync(string parentTenantId)
...
                 child.Id = parentTenantId.CreateChildId();
 
-                // At the least, we need to copy the default blob storage settings from its parent
+                // At the least, we need to copy the tenant blob storage settings from its parent
                 // to support the tenant blob store provider. We would expect this to be overridden
                 // by clients that wanted to establish their own settings.
-                BlobStorageConfiguration? defaultStorageConfiguration = parentTenant.GetDefaultBlobStorageConfiguration();
-                child.SetDefaultBlobStorageConfiguration(defaultStorageConfiguration!);
-                if (parentTenant.HasStorageBlobConfiguration(this.ContainerDefinition))
+                BlobStorageConfiguration? defaultStorageConfiguration = parentTenant.GetBlobStorageConfiguration(ContainerDefinition);
```
Comment (idg10) [inline Solutions/Corvus.Tenancy.Storage.Azure.Blob/Corvus/Tenancy/Storage/Azure/TenantProviderBlobStore.cs]: I think `defaultStorageConfiguration` might no longer be the correct name for this variable.

File: Solutions/Corvus.Tenancy.Storage.Azure.Blob/Corvus/Tenancy/Storage/Azure/TenantProviderBlobStore.cs [hunk 4]
```diff
@@ -166,14 +166,14 @@ public async Task<ITenant> CreateChildTenantAsync(string parentTenantId)
                 Tenant child = this.serviceProvider.GetRequiredService<Tenant>();
                 child.Id = parentTenantId.CreateChildId();
 
-                // At the least, we need to copy the default blob storage settings from its parent
+                // At the least, we need to copy the tenant blob storage settings from its parent
```
Comment (idg10) [inline Solutions/Corvus.Tenancy.Storage.Azure.Blob/Corvus/Tenancy/Storage/Azure/TenantProviderBlobStore.cs]: I don't know if this change to the comments is quite sufficient.

I think I'd prefer it to state explicitly that we're only copying the tenant blob storage settings for the `corvustenancy` container (i.e., the one described by `ContainerDefinition`). (And this is a change, right. Before, this was actually copying non-container-specific default blob storage settings.)

Second, I'm not sure that it's really true that we "would expect this to be overridden by clients that wanted to establish their own settings". We have the ability for clients to be able to specify that they want to use a different storage account for all their child tenants, but I think most of the time we're going to expect clients just to use the same storage account as everyone else for their tenant settings. (I'm not even sure, given our newly defined approach to tenancy settings, whether we really even offer client-tenancy-settings isolation: we can store a client's child tenant settings in its own storage account, but all of the corresponding on-behalf-of tenants actually live under the relevant service tenants, which will not be using client-specific storage. So the reality is that even if a customer brings their own storage for tenancy, we'll still end up copying their settings into the common storage account any time we create an on-behalf-of tenant. So I think this idea of tenant-specific storage for tenancy settings may in effect be a fiction. This makes me wonder whether we should actually dramatically simplify this whole thing and say that the `TenantProviderBlobStore` just uses the storage account it's told to, and to remove all the code that walks through the tenant tree working out which store to use for any particular child tenant.)

File: Solutions/Corvus.Azure.Storage.Tenancy/Microsoft/Extensions/DependencyInjection/TenancyBlobStorageServiceCollectionExtensions.cs (hunk 1 above)
Comment (jongeorge1) [reply Solutions/Corvus.Azure.Storage.Tenancy/Microsoft/Extensions/DependencyInjection/TenancyBlobStorageServiceCollectionExtensions.cs]: Weird.

File: Solutions/Corvus.Azure.Storage.Tenancy/Microsoft/Extensions/DependencyInjection/TenancyBlobStorageServiceCollectionExtensions.cs (hunk 2 above)
Comment (jongeorge1) [reply Solutions/Corvus.Azure.Storage.Tenancy/Microsoft/Extensions/DependencyInjection/TenancyBlobStorageServiceCollectionExtensions.cs]: Yes, I had intended to remove these but only got rid of one.

File: Solutions/Corvus.Tenancy.Storage.Azure.Blob/Corvus/Tenancy/Storage/Azure/TenantProviderBlobStore.cs (hunk 4 above)
Comment (jongeorge1) [reply Solutions/Corvus.Tenancy.Storage.Azure.Blob/Corvus/Tenancy/Storage/Azure/TenantProviderBlobStore.cs]: I've rejigged the comment. As discussed, the larger point will be dealt with via a separate discussion/issue.

File: Solutions/Corvus.Azure.Storage.Tenancy/Microsoft/Extensions/DependencyInjection/TenancyBlobStorageServiceCollectionExtensions.cs (hunk 2 above)
Comment (jongeorge1) [reply Solutions/Corvus.Azure.Storage.Tenancy/Microsoft/Extensions/DependencyInjection/TenancyBlobStorageServiceCollectionExtensions.cs]: Yeah, I should have removed that.

**Pull request merged**